all-sdk/autotuning.md

```diff
@@ -0,0 +1,431 @@
+
+  /// Frees a slot.
+  fn release_slot(&self, info: &Self::SlotKind::Info);
+  
+  /// If this implementation knows how many slots are available at any moment, it should return that here.
+  fn available_slots(&self) -> Option<usize>;
+}
+
+// These functions would be provided to the user as top level statics they can access to get information about how
+// all outstanding slots are currently being used. Ostensibly they could've stored this in their own implementation
+// too, but this is to make it convenient for them to answer questions like "How many slots are being used by 
+// activities of type X right now?". 
+//
+// These could potentially be provided as parameters to the trait methods - but at least in the case of the 
+// async/blocking reserve, the information may change over time, so it may be best to have these be explicit calls.
+pub fn used_workflow_slots_info() -> WorkflowSlotsInfo { /* ... */ }
+pub fn used_activity_slots_info() -> ActivitySlotsInfo { /* ... */ }
+pub fn used_local_activity_slots_info() -> LocalActivitySlotsInfo { /* ... */ }
+
+struct WorkflowSlotsInfo {
+  used_slots: Vec<WorkflowSlotInfo>,
+  /// Current size of the workflow cache.
+  num_cached_workflows: usize,
+  /// The limit on the size of the cache, if any. This is important for users to know as discussed below in the section
+  /// on workflow cache management.
+  max_cache_size: Option<usize>,
+  // ... Possibly also metric information
+}
+struct ActivitySlotsInfo {
+  used_slots: Vec<ActivitySlotInfo>,
+}
+struct LocalActivitySlotsInfo {
+  used_slots: Vec<LocalActivitySlotInfo>,
+}
+
+struct WorkflowSlotInfo {
+  workflow_type: String,
+  // task queue, worker id, etc...
+}
+struct ActivitySlotInfo {
+  activity_type: String,
+  // etc...
+}
+struct LocalActivitySlotInfo {
+  activity_type: String,
+  // etc...
+}
+
+struct WorkflowSlotKind {}
+struct ActivitySlotKind {}
+struct LocalActivitySlotKind {}
+trait SlotKind {
+  type Info;
+}
+impl SlotKind for WorkflowSlotKind {
+  type Info = WorkflowSlotInfo;
+}
+impl SlotKind for ActivitySlotKind {
+  type Info = ActivitySlotInfo;
+}
+impl SlotKind for LocalActivitySlotKind {
+  type Info = LocalActivitySlotInfo;
+}
+trait WorkflowTaskSlotSupplier: SlotSupplier<SlotKind=WorkflowSlotKind> {}
+trait ActivityTaskSlotSupplier: SlotSupplier<SlotKind=ActivitySlotKind> {}
+trait LocalActivityTaskSlotSupplier: SlotSupplier<SlotKind=LocalActivitySlotKind> {}
+
+/// Users might want to be able to pause the handing-out of slots as an effective way of pausing their workers.
+/// We can provide an implementation for this that wraps their implementation, or one of the defaults we provide.
+struct PauseableSlotSupplier<T: SlotKind> {
+  inner: Arc<dyn SlotSupplier<SlotKind=T>>,
+  paused: AtomicBool,
+  // ... details
+}
+impl<T: SlotKind> PauseableSlotSupplier<T> {
+  pub fn new(supplier: impl SlotSupplier<SlotKind=T>) -> Self { /* ... */ }
+  pub fn pause(&self) { /* ... */ }
```

Comment:
There is a interesting question on when `pause` returns. Do you return when you stop handing out new slots or return when you have no more slots in "inflight" ie reserved but not marked in use

Reply:
IMO it should return immediately, pause is a state not an actionable request for the most part.

Reply:
Yeah, i am imagining it returns immediately. You're saying "become paused" (like initiating shutdown of a worker)